src/mame/machine/cxd8452aq.cpp

```diff
@@ -0,0 +1,281 @@
+	m_sonic3_reg.tx_host_address = data;
+}
+
+uint32_t cxd8452aq_device::tx_count_r(offs_t offset)
+{
+	return m_sonic3_reg.tx_count;
+}
+
+void cxd8452aq_device::tx_count_w(offs_t offset, uint32_t data)
+{
+	LOGMASKED(LOG_REGISTERS, "set sonic3.tx_count = 0x%x\n", data);
+
+	m_sonic3_reg.tx_count = data;
+	if (data & DMA_START)
+	{
+		m_dma_check->adjust(attotime::zero);
+	}
+	m_irq_check->adjust(attotime::zero);
+}
+
+uint32_t cxd8452aq_device::rx_count_r(offs_t offset)
+{
+	return m_sonic3_reg.rx_count;
+}
+
+void cxd8452aq_device::rx_count_w(offs_t offset, uint32_t data)
+{
+	LOGMASKED(LOG_REGISTERS, "set sonic3.rx_count = 0x%x\n", data);
+
+	m_sonic3_reg.rx_count = data;
+	if (data & DMA_START)
+	{
+		m_dma_check->adjust(attotime::zero);
+	}
+	m_irq_check->adjust(attotime::zero);
+}
+
+void cxd8452aq_device::sonic_bus_map(address_map &map)
+{
+	map(0x0, 0xfffff).mirror(0xfff00000).rw(FUNC(cxd8452aq_device::sonic_r), FUNC(cxd8452aq_device::sonic_w));
+}
+
+uint8_t cxd8452aq_device::sonic_r(offs_t offset)
+{
+	return space(0).read_byte(offset);
+}
+
+void cxd8452aq_device::sonic_w(offs_t offset, uint8_t data)
+{
+	space(0).write_byte(offset, data);
+}
+
+TIMER_CALLBACK_MEMBER(cxd8452aq_device::irq_check)
+{
+	bool newIrq = m_sonic3_reg.control & (EXT_INT | RX_DMA_COMPLETE | TX_DMA_COMPLETE);
+	if (m_irq != newIrq)
+	{
+		LOGMASKED(LOG_INTERRUPT, "sonic3 interrupt changed to %d!\n", newIrq);
+		m_irq = newIrq;
+		m_irq_handler(newIrq);
+	}
+}
+
+TIMER_CALLBACK_MEMBER(cxd8452aq_device::dma_check)
+{
+	bool rxDmaActive = m_sonic3_reg.rx_count > DMA_START;
+	bool txDmaActive = m_sonic3_reg.tx_count > DMA_START;
+
+	if (rxDmaActive && txDmaActive)
+	{
+		// TODO: is this an error? allow it for now, but log it.
+		LOGMASKED(LOG_GENERAL, "WARNING: RX and TX DMA active at the same time, this may not be a valid condition.\n");
+	}
+
+	if (rxDmaActive)
+	{
+		// Move byte from SONIC RAM to main memory
+		uint8_t data = space(0).read_byte(m_sonic3_reg.rx_sonic_address & 0xffff);
+		LOGMASKED(LOG_DATA, "sonic3.rx(0x%x -> 0x%x, data=0x%x)\n", m_sonic3_reg.rx_sonic_address, m_sonic3_reg.rx_host_address, data);
+		m_bus->write_byte(m_apbus_virt_to_phys_callback(m_sonic3_reg.rx_host_address), data);
+		m_sonic3_reg.rx_count -= 1;
+		m_sonic3_reg.rx_sonic_address += 1;
+		m_sonic3_reg.rx_host_address += 1;
+		if (m_sonic3_reg.rx_count == DMA_START)
+		{
+			m_sonic3_reg.control |= RX_DMA_COMPLETE;
+
+			// TODO: determine if rx_count is cleared here
+			m_sonic3_reg.rx_count = 0;
+			m_irq_check->adjust(attotime::zero);
+		}
+	}
+
+	if (txDmaActive)
+	{
+		// Move byte from main memory to SONIC RAM
+		uint8_t data = m_bus->read_byte(m_apbus_virt_to_phys_callback(m_sonic3_reg.tx_host_address));
+		LOGMASKED(LOG_DATA, "sonic3.tx(0x%x -> 0x%x, data=0x%x)\n", m_sonic3_reg.tx_host_address, m_sonic3_reg.tx_sonic_address, data);
+		space(0).write_byte(m_sonic3_reg.tx_sonic_address & 0xffff, data);
```

Comment:
You probably want to use one of the accessor cache things for this to save time on frequent accesses (a `memory_access_specific` or `memory_access_cache` member).

Reply:
Done!